src/client/core/aad/auth-provider.ts

```diff
@@ -0,0 +1,113 @@
+import {
+    AccountInfo,
+    AuthenticationResult,
+    ClientApplication,
+    PublicClientApplication
+} from "@azure/msal-node";
+import { BatchExplorerApplication } from "..";
+import { AADConfig } from "./aad-config";
+
+const MSAL_SCOPES = ["user_impersonation"];
+
+export type AuthorizationResult = AuthenticationResult;
+
+/**
+ * Provides authentication services
+ */
+export default class AuthProvider {
+    private _clients: StringMap<ClientApplication> = {};
+    private _accounts: StringMap<AccountInfo> = {};
+
+    constructor(
+        protected app: BatchExplorerApplication,
+        protected config: AADConfig
+    ) {}
+
+    /**
+     * Retrieves an access token
+     *
+     * Will attempt to retrieve the token silently if an account exists in the
+     * cache, but will fall back to interactive access token retrieval.
+     *
+     * @param authCodeCallback Handles interactive authentication code retrieval
+     */
+    public async getToken(options: {
+        resourceURI: string,
+        tenantId?: string,
+        authCodeCallback: (url: string) => Promise<string>
+    }): Promise<AuthorizationResult> {
+        const { resourceURI, tenantId = "common", authCodeCallback } = options;
+
+        /**
+         * KLUDGE: msal.js does not handle well access tokens across multiple
+         * tenants within the same cache. It lets you specify a different
+         * authority per request but it returns the same access token.
+         *
+         * Until this is resolved, we use one client application per tenant.
+         */
+        const client = this._getClient(tenantId);
+
+        const authRequest = this._authRequest(resourceURI, tenantId);
+        if (this._accounts[tenantId]) {
+            const result = await client.acquireTokenSilent({
+                ...authRequest, account: this._accounts[tenantId]
+            });
+            return result;
+        } else {
+            const url = await client.getAuthCodeUrl(authRequest);
+            const code = await authCodeCallback(url);
+            const result: AuthorizationResult =
+                await client.acquireTokenByCode({ ...authRequest, code });
+            if (result) {
+                this._accounts[tenantId] = result.account;
+            }
+            return result;
+        }
+    }
+
+    public logout(): void {
+        this._removeAccount();
```

Comment:
logout removes the client and account from their respective caches, but is it possible something could be making calls using an instance previously fetched after the user thought they were logged out?

Reply:
I'm not sure I totally understand what you mean by "instance previously fetched", but `AuthProvider.logout()` simply resets its own conception of a logged-in user, and cannot ensure that the credentials are removed from the entire system (e.g., browser cache). Subsequent access token fetches will require interactive authentication. The authentication service also clears browser cookies when calling `AuthProvider.logout()` ([relevant lines](https://github.com/Azure/BatchExplorer/blob/9a037200750fd27cb0d8931e97fca26c64c29d99/src/client/core/aad/authentication/authentication.service.ts#L119-L120))